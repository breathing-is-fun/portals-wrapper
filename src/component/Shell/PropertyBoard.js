/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-10 10:12:51
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-10 16:15:23
 */
import React, { Component } from 'react';

import { Drawer, Icon, Form, Input } from 'antd';
const FormItem = Form.Item;

class PropertyBoard extends Component {
	constructor (props) {
		super(props);

		this.state = {

		};
	}

    componentDidMount = () => {

    }

	handleSubmit = e => {
		e.preventDefault();

		this.props.form.validateFieldsAndScroll((err, values) => {
			if (!err) {
			  	console.log('Received values of form: ', values);
			}
		});
	}

    render = () => {
    	const { visible, onClose, dataSource, form } = this.props;
    	const { style, title } = dataSource;
    	const { getFieldDecorator  } = form;

    	const drawerTitle = (
    		<div>
    			<Icon type='form' theme='outlined' />
    			<span style={{ paddingLeft: 10 }}>{ `${ title } 外壳属性修改` }</span>
    		</div>
    	);

    	const drawerProps = {
    		visible,
    		className: 'PropertyBoard',
    		onClose: () => onClose && onClose(!visible),
    		title: drawerTitle,
    		width: '20%',
    	};

    	const formItemLayoutProps = {
    		labelCol: {
    			xs: { span: 24 },
    			sm: { span: 4 },
    		},
    		wrapperCol: {
    			xs: { span: 24 },
    			sm: { span: 18 },
    		},
    	};

    	return (
    		<div className='PropertyBoard'>
    			<Drawer { ...drawerProps }>
    				<Form onSubmit={ this.handleSubmit }>
    					<FormItem { ...formItemLayoutProps } label='标题'>
    						{
    							getFieldDecorator('email', {
    								rules: [{
    									type: 'email', message: 'The input is not valid E-mail!',
    								}, {
    									required: true, message: 'Please input your E-mail!',
    								}],
    							})(<Input />)
    						}
    					</FormItem>
    				</Form>
    			</Drawer>
    		</div>
    	);
    }
}

export default Form.create()(PropertyBoard);