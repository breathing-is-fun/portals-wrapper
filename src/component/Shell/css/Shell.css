.Shell .operation {
    cursor: pointer;
    padding: 5px;
    float: right;
    position: absolute;
    top: 0px;
}

.Shell .operation-title {
    font-family: MicrosoftYaHei;
    font-size: 14px;
    font-weight: normal;
    font-stretch: normal;
    letter-spacing: 1px;
    color: #171717;
    padding-left: 10px;
}

.Shell {
    transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
    z-index: 1;
    user-select: none;
    box-shadow: 0px 0px 29px 0px rgba(93, 106, 113, 0.12);
    border-radius: 2;
}

.Shell:hover {
    transform: scale(1.05);
}

.border-transition:hover {
    border: 1px dashed #1890ff;
}

.border-transition {
    border: 1px dashed #d9d9d9;
}