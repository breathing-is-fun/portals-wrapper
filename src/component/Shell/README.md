English | [简体中文](./README-zh_CN.md)

# 简介

- [格子](../Grid/README-zh_CN.md)的外壳

# API  

| 名称 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| data-grid | 这个参数是 `react-grid-layout` 要求的，没这个就没法布局，[了解一下](https://github.com/STRML/react-grid-layout#grid-item-props)。后续会修改 | {} | - |
| showDetail | 是否显示格子右上的详情按钮，详情以全局弹框的方式引入 | boolean | false |
| detailPath | 详情页地址，暂时只支持 `iframe` 嵌入 | string | - |
| onDetail | 点击详情按钮的回调 | function (item: data-grid) | - |
| showEdit | 是否显示右上编辑按钮 | boolean | false |
| onEdit | 点击编辑按钮回调 | function (item: data-grid) | - |
| showDelete | 是否显示右上删除按钮 | boolean | false |
| onDelete | 点击删除按钮回调 | function (data-grid) | - |
| type | 外壳类型，用作判断是谁用了 `Shell`，可以用作权限控制 | string, enum['add', 'module', 'component'] | 'component' |
| onAdd | [套餐](http://www.baidu.com)编辑页点击添加按钮的回调 | function (e: MouseEvent) | - |
| showTitle | 是否显示标题 | boolean | false |
| style | 控制最外层样式 | {} | - |
| title | 标题 | `string | ReactNode` | - |

# 示例

``` jsx
    import Shell from '.'

    render => <Shell>test</Shell>
```