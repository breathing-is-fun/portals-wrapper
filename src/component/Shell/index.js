/*
 * @Author: zy9@github.com/zy410419243
 * @Date: 2018-10-08 10:39:22
 * @Last Modified by: zy9
 * @Last Modified time: 2018-10-11 11:06:31
 */
import React, { Component } from 'react';

import { Icon } from 'antd';
// import PropertyBoard from '../PropertyBoard';

import './css/Shell.css';

export default class Shell extends Component {
	constructor (props) {
		super(props);
		const { title } = props;

		this.state = {
			title,
			propertyBoardVisible: false,
			propertyBoardDataSource: {}
		};
	}

    componentDidMount = () => {

    }

	handleOnDelete = e => {
		const { onDelete } = this.props;
		const dataGrid = this.props['data-grid'] || {};

		this.preventBubble(e);

		onDelete && onDelete(dataGrid);
	}

	handleOnEdit = e => {
		const { style, onEdit } = this.props;
		const dataGrid = this.props['data-grid'] || {};

		this.preventBubble(e);

		this.setState({
			propertyBoardVisible: true,
			propertyBoardDataSource: { title: dataGrid.title || '', style },
		}, () => {
			onEdit && onEdit(true);
		});
	}

	// 阻止react-grid-layout的拖拽事件在点击时生效
	preventBubble = e => {
		if(e.stopPropagation) {
			e.stopPropagation();
		} else {
			e.cancelBubble = true;
		}
	}

	handleDrawerOnClose = propertyBoardVisible => {
		const { onEdit } = this.props;

		this.setState({ propertyBoardVisible }, () => {
			onEdit && onEdit(propertyBoardVisible);
		});
	}

    render = () => {
    	const { children } = this.props;
    	const { title, propertyBoardVisible, propertyBoardDataSource } = this.state;

    	const operateBoard = (
    		<div style={{ background: '#ccc', height: 30 }}>
    			<Icon type='delete' theme='outlined' className='operation' onMouseDown={ this.handleOnDelete } />
    			<Icon type='edit' theme='outlined' className='operation' onMouseDown={ this.handleOnEdit } />
    			<span>{ title }</span>
    		</div>
    	);

    	const propertyBoardProps = {
    		visible: propertyBoardVisible,
    		onClose: this.handleDrawerOnClose,
    		dataSource: propertyBoardDataSource,
    	};

    	// 容错，getDrawerStatus为react不识别的handle
    	let newProps = Object.assign({}, this.props);

    	delete newProps.getDrawerStatus;
    	delete newProps.onDelete;
    	delete newProps.onEdit;

    	return (
    		<div { ...newProps } className='Shell'>
    			{ operateBoard }

    			{ children }

    			{/* <PropertyBoard { ...propertyBoardProps } /> */}
    		</div>
    	);
    }
}