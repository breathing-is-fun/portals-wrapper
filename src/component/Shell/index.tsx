import React, { Component } from 'react';
import classNames from 'classnames';
import { Icon } from 'antd';
import omit from 'omit.js';
import './css/Shell.css';

export interface ShellProps {
  showDetail: boolean;
  detailPath: string;
  showEdit: boolean;
  showDelete: boolean;
  type?: 'add' | 'module' | 'component' | string;
  showTitle: boolean;
  style: any;
  title: any;
  'data-grid': any;
  size?: string | number;
  onAdd?: (dataGrid: any) => void;
  onDelete?: (dataGrid: any) => void;
  onDetail?: (dataGrid: any) => void;
  onEdit?: (dataGrid: any) => void;
}

export default class Shell extends Component<ShellProps> {
  static defaultProps = {
    showEdit: false,
    showDelete: false,
    showDetail: false,
    detailPath: '',
    type: 'component',
    showTitle: true,
    style: {},
    title: '',
    'data-grid': {},
  };

  handleOnDelete = (e: any) => {
    const { onDelete } = this.props;
    const dataGrid = this.props['data-grid'] || {};

    this.preventBubble(e);

    onDelete && onDelete(dataGrid);
  };

  handleOnEdit = (e: any) => {
    const { onEdit } = this.props;
    const dataGrid = this.props['data-grid'] || {};

    this.preventBubble(e);

    onEdit && onEdit(dataGrid);
  };

  handleOnDetail = (e: any) => {
    const { onDetail } = this.props;
    const dataGrid = this.props['data-grid'] || {};

    this.preventBubble(e);

    onDetail && onDetail(dataGrid);
  };

  // 阻止react-grid-layout的拖拽事件在点击时生效
  preventBubble = (e: any) => {
    if (e.stopPropagation) {
      e.stopPropagation();
    } else {
      e.cancelBubble = true;
    }
  };

  render = () => {
    const {
      children,
      showEdit,
      showDelete,
      type,
      onAdd,
      showTitle,
      style,
      showDetail,
      title,
      size,
    } = this.props;

    const editButton = (
      <Icon
        type="edit"
        className="operation"
        onMouseDown={this.handleOnEdit}
        style={{ right: 40 }}
      />
    );
    const deleteButton = (
      <Icon
        type="delete"
        className="operation"
        onMouseDown={this.handleOnDelete}
        style={{ right: 10 }}
      />
    );
    // const detailButton = (
    //   <Icon
    //     type='small-dash'
    //     className='operation'
    //     onMouseDown={this.handleOnDetail}
    //     style={{ right: 10, fontSize: '2vw' }}
    //   />
    // );
    const detailButton = (
      <span
        onMouseDown={this.handleOnDetail}
        style={{
          right: 10,
          position: 'absolute',
          cursor: 'pointer',
          fontSize: 17,
          top: 6,
        }}
      >
        详情 >
      </span>
    );

    const newProps = omit(this.props, [
      'showDelete',
      'onDelete',
      'showDetail',
      'onEdit',
      'showEdit',
      'title',
      'showTitle',
      'onDetail',
      'detailPath',
      'onAdd',
    ]);

    if ((window as any).SCTool) {
      SCTool.listener.set(newProps['data-grid'].i, {
        width: newProps.style.width,
        height: newProps.style.height,
      });
    }

    return (
      <div
        {...newProps}
        style={Object.assign(
          {},
          {
            overflow: showEdit ? 'hidden' : 'auto',
          },
          style,
        )}
        className={classNames({
          Shell: true,
          'border-transition': type != 'component',
        })}
        onClick={e => type == 'add' && onAdd && onAdd(e)}
      >
        {showDelete && deleteButton}

        {showEdit && editButton}

        {showDetail && detailButton}

        {type != 'add' && showTitle && (
          <span
            className={classNames({
              'sc-title': size == '',
              'sc-title-sm': size == 'sm',
              'operation-title': true,
            })}
          >
            {title}
          </span>
        )}

        {children}
      </div>
    );
  };
}
