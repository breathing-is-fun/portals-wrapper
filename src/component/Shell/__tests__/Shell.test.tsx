import React from 'react';
import { mount, shallow } from 'enzyme';
import Shell from '..';

describe('Shell', () => {
  const props = {
    'data-grid': {
      h: 6,
      i: 'dbzx',
      imgUrl: 'http://47.95.1.229:9003/UploadFile/201808/dbzx.png',
      path: '../../../thirdModules/dbzx/index.js',
      static: false,
      style: {},
      title: 'testTitle',
      w: 4,
      x: 0,
      y: 0,
    },
    showDelete: true,
    showEdit: true,
    key: 'testKey',
    onDelete: jest.fn(),
    onEdit: jest.fn(),
    style: { zIndex: 1, background: '#F96' },
    title: 'testTitle',
    onAdd: jest.fn(),
  };

  it('render correctly', () => {
    const wrapper = shallow(<Shell {...props}>test</Shell>);

    expect(wrapper).toMatchSnapshot();
  });

  it('button group render correctly', () => {
    const wrapper = shallow(<Shell {...props}>test</Shell>);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ isDelete: false, isEdit: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('when button is clicked, theirs click event should be called', () => {
    const wrapper = mount(<Shell {...props}>test</Shell>);

    wrapper
      .find('i')
      .at(0)
      .simulate('mousedown');
    expect(props.onDelete.mock.calls.length).toBe(1);

    wrapper
      .find('i')
      .at(1)
      .simulate('mousedown');
    expect(props.onEdit.mock.calls.length).toBe(1);
  });

  it('when showTitle is false, title should be hidden', () => {
    const demo = (
      <Shell {...Object.assign({}, props, { showTitle: false })}>test</Shell>
    );
    const wrapper = mount(demo);

    expect(wrapper.find('.operation-title').length).toBe(0);
  });

  it('when add button is clicked, its click event should be called', () => {
    const demo = (
      <Shell {...Object.assign({}, props, { type: 'add' })}>test</Shell>
    );
    const wrapper = shallow(demo);

    wrapper.find('.Shell').simulate('click');
    expect(props.onAdd.mock.calls.length).toBe(1);
  });
});
