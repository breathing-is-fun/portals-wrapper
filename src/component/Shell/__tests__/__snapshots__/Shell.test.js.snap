// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shell button group render correctly 1`] = `
<div
  className="Shell"
  data-grid={
    Object {
      "h": 6,
      "i": "dbzx",
      "imgUrl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
      "path": "../../../thirdModules/dbzx/index.js",
      "static": false,
      "style": Object {},
      "title": "testTitle",
      "w": 4,
      "x": 0,
      "y": 0,
    }
  }
  onClick={[Function]}
  style={
    Object {
      "background": "#F96",
      "zIndex": 1,
    }
  }
  type="component"
>
  <Icon
    className="operation"
    onMouseDown={[Function]}
    style={
      Object {
        "right": 10,
      }
    }
    theme="outlined"
    type="delete"
  />
  <Icon
    className="operation"
    onMouseDown={[Function]}
    style={
      Object {
        "right": 40,
      }
    }
    theme="outlined"
    type="edit"
  />
  <span
    className="operation-title"
  >
    testTitle
  </span>
  test
</div>
`;

exports[`Shell button group render correctly 2`] = `
<div
  className="Shell"
  data-grid={
    Object {
      "h": 6,
      "i": "dbzx",
      "imgUrl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
      "path": "../../../thirdModules/dbzx/index.js",
      "static": false,
      "style": Object {},
      "title": "testTitle",
      "w": 4,
      "x": 0,
      "y": 0,
    }
  }
  onClick={[Function]}
  style={
    Object {
      "background": "#F96",
      "zIndex": 1,
    }
  }
  type="component"
>
  <span
    className="operation-title"
  >
    testTitle
  </span>
  test
</div>
`;

exports[`Shell render correctly 1`] = `
<div
  className="Shell"
  data-grid={
    Object {
      "h": 6,
      "i": "dbzx",
      "imgUrl": "http://47.95.1.229:9003/UploadFile/201808/dbzx.png",
      "path": "../../../thirdModules/dbzx/index.js",
      "static": false,
      "style": Object {},
      "title": "testTitle",
      "w": 4,
      "x": 0,
      "y": 0,
    }
  }
  onClick={[Function]}
  style={
    Object {
      "background": "#F96",
      "zIndex": 1,
    }
  }
  type="component"
>
  <Icon
    className="operation"
    onMouseDown={[Function]}
    style={
      Object {
        "right": 10,
      }
    }
    theme="outlined"
    type="delete"
  />
  <Icon
    className="operation"
    onMouseDown={[Function]}
    style={
      Object {
        "right": 40,
      }
    }
    theme="outlined"
    type="edit"
  />
  <span
    className="operation-title"
  >
    testTitle
  </span>
  test
</div>
`;
